java-squid/src/main/java/org/sonar/java/se/SymbolicValue.java

```diff
@@ -0,0 +1,415 @@
+      }
+      return ImmutableList.of(programState);
+    }
+
+    @Override
+    public List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
+      Object data = programState.constraints.get(this);
+      // update program state only for a different constraint
+      if (data instanceof BooleanConstraint) {
+        BooleanConstraint bc = (BooleanConstraint) data;
+        if ((BooleanConstraint.TRUE.equals(booleanConstraint) && BooleanConstraint.FALSE.equals(bc)) ||
+          (BooleanConstraint.TRUE.equals(bc) && BooleanConstraint.FALSE.equals(booleanConstraint))) {
+          // setting null where value is known to be non null or the contrary
+          return ImmutableList.of();
+        }
+      }
+      if (data == null || !data.equals(booleanConstraint)) {
+
+        // store constraint only if symbolic value can be reached by a symbol.
+        if (programState.values.containsValue(this)) {
+          Map<SymbolicValue, Object> temp = Maps.newHashMap(programState.constraints);
+          temp.put(this, booleanConstraint);
+          return ImmutableList.of(new ProgramState(programState.values, temp, programState.visitedPoints, programState.stack));
+        }
+      }
+      return ImmutableList.of(programState);
+    }
+
+    @Override
+    public ProgramState setSingleConstraint(ProgramState programState, NullConstraint nullConstraint) {
+      final List<ProgramState> states = setConstraint(programState, nullConstraint);
+      if (states.size() != 1) {
+        throw new IllegalStateException("Only a single program state is expected at this location");
+      }
+      return states.get(0);
+    }
+  }
+
+  abstract class BinarySymbolicValue extends ObjectSymbolicValue {
+
+    SymbolicValue leftOp;
+    SymbolicValue rightOp;
+
+    public BinarySymbolicValue(int id) {
+      super(id);
+    }
+
+    abstract BooleanConstraint shouldNotInverse();
+
+    @Override
+    public void computedFrom(List<SymbolicValue> symbolicValues) {
+      Preconditions.checkArgument(symbolicValues.size() == 2);
+      rightOp = symbolicValues.get(0);
+      leftOp = symbolicValues.get(1);
+    }
+
+    @Override
+    public List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
+      if (leftOp.equals(rightOp)) {
+        if (shouldNotInverse().equals(booleanConstraint)) {
+          return ImmutableList.of(programState);
+        }
+        return ImmutableList.of();
+      }
+      List<ProgramState> result = copyConstraint(leftOp, rightOp, programState, booleanConstraint);
+      if (result.isEmpty()) {
+        return ImmutableList.of();
+      }
+      List<ProgramState> reversedResult = new ArrayList<>();
+      for (ProgramState ps : result) {
+        List<ProgramState> reversedStates = copyConstraint(rightOp, leftOp, ps, booleanConstraint);
+        if (reversedStates.size() == 1 && reversedStates.get(0) == programState) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)